### ISSUE TITLE
Delay/Synchronization of measurements or channels in eCAL Player

### ISSUE THREAD

**Issue by nikcleju** [continental/ecal#336]
I need to synchronize two separate eCAL measurements, i.e. play both of them simultaneously but with a certain delay added to one of them such that one data flow arrives at the subscriber in sync with the other one. The delay needs manually adjustments a few times until I figure out the right value. 

What is the best solution to achieve this? 

I'm thinking of 3 ways of doing this, but I don't know which is better:

1. Play them with two separate Player instances, and use a bash script or eCAL Sys to start one of them after a delay.  In this case I worry about extra delays when starting the play, which may ruin the synchronization. Should I expect delays if I use interactive CLI mode,  load the measurements beforehand, then simply pass a space character to start the players?

2. Make a single eCAL measurement with all the channels (play them both and rerecord), the hack around in MeasurementContainer::CreateFrameTable() and add a fixed delay to the receive_timestamp_ and send_timestamp_ of certain channels. I'm not sure if this won't break other things in the code, but it seems to me a flexible solution of adding configurable delays to the individual channels.

3.  Write a custom Python script that subscribes to all the channels, receives all the data and publishes it back with configurable delays.

Thanks for any hint!

**Comment by FlorianReimold** [comment]
OK, so the proper way would be to modify the timestamps of those measurements and then copy them into one directory, so the Player will load the data as one continuous measurement. Unfortunatelly we don't have a ready-to-use solution at the moment.

1. I agree, there may be delays that you cannot control, like the load time of the measurement itself. If you have already loaded the measurement and connected the player to eCAL you can minimize that delay.
2.  Well, I guess that should work, but keep in mind that every time you play and record, you may loose information (e.g. time between frames, dropped frames etc). Sounds like a very ugly hack though and you would have to recompile ecal every time.
3. I guess that should be possible, but if you have a high data rate, you may e.g. drop frames or otherwise decrease the quality of your measurement


If you are willing to do some coding, I would suggest to write an application that loads the HDF5 files, modifies the timestamps and saves a new HDF5 file. You could use the eCAL Measurement API for that.
https://github.com/continental/ecal/tree/master/samples/cpp/measurement

**Comment by nikcleju** [comment]
Thanks for the answer! 

I agree that the best way is to write a dedicated tool to modify the timestamps in a HDF5 file, but only after I know exactly by what amount.

For quicker experimentation with delays, I went along with solution 2: 
- drop all HDF5 files in a single directory, to have a continuous measurement
- make a file `/doc/delays.txt` with channel names and associated delays, which are read and added to the timestamps when the measurement is loaded in MeasurementContainer::CreateFrameTable(). Then I only need to change the values in the txt file and reload the measurements. It seems to be working OK, based on an old 5.4.1 version we're using.
- I'm currently working to have an editable delay column in the GUI app as well, which would trigger MeasurementContainer::CreateFrameTable() and basically reload the measurements to apply the new delays.

**Comment by rex-schilasky** [comment]
Hi,

as a side node. Internally we use an additional eCAL eco system tool (eCALConv) that is unfortunately not part of the open source repository for today. This tool can convert HDF5 into other formats and by a plugin architecture shift timestamps if needed. We hope we can add this small utility later.

But anyway .. you really should upgrade your eCAL version, there are so many improvements and bugfixes done the last 22 month ;-).

**Comment by nikcleju** [comment]
Thanks for the info. I will look into eCALConv, since I have access to the internal repository (I'm with Continental, working for an internal project).

**Comment by rex-schilasky** [comment]
Ahhh .. that is perfect. Contact me by mail if you need further information.
